### PR TITLE
add azurerm_subnet_network_security_group_association

### PR DESCRIPTION
**Pull request by ar3ndt**

### CODE REVIEW

File: resources/terraform/networking.tf [hunk 1]
```diff
@@ -14,8 +14,8 @@ resource "azurerm_subnet" "subnets" {
 }
 
 resource "azurerm_network_security_group" "nsg" {
-  count               = local.create_public_access_nsg == true ? 1 : 0
-  name                = "${var.name}-nsg"
+  count               = length(var.subnets)
```
Comment (atsikham) [inline resources/terraform/networking.tf:17]: Do we need to have such amount of nsg? Can 1 nsg be assigned to multiple subnets?

**Review by atsikham — CHANGES_REQUESTED** [review]
Besides requested changes please check why pipeline fails.

**Comment by mkyc** [comment]
Also please notice that tests are failing.

File: resources/terraform/locals.tf [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
-}
+# locals {
+#   create_public_access_nsg = length(
+#     [
+#       for vm_group in var.vm_groups:
+#       vm_group.use_public_ip
+#       if vm_group.use_public_ip == true
+#     ]
+#   ) > 0
+# }
```
[the remark this answers is not in the document]
Comment (ar3ndt) [reply resources/terraform/locals.tf:9]: it actually will not be commented once we use only 1 nsg

File: resources/terraform/networking.tf (hunk 1 above)
Comment (ar3ndt) [reply resources/terraform/networking.tf:17]: yup it can. I am changing this.

**Comment by ar3ndt** [comment]
> Besides requested changes please check why pipeline fails.

fixed

**Review by atsikham — APPROVED** [review]

**Review by mkyc — CHANGES_REQUESTED** [review]
When running plan command with `--debug` flag you'll see following warning. It's worth to clear it up: 

...

> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Plan: 11 to add, 0 to change, 0 to destroy.
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Warning: Interpolation-only expressions are deprecated
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >   on networking.tf line 37, in resource "azurerm_subnet_network_security_group_association" "subnets_to_nsg":
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >   37:   subnet_id                 = "${azurerm_subnet.subnets[count.index].id}"
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Terraform 0.11 and earlier required all non-constant expressions to be
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > provided via interpolation syntax, but this pattern is now deprecated. To
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > silence this warning, remove the "${ sequence from the start and the }"
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > sequence from the end of this expression, leaving just the inner expression.
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Template interpolation syntax is still used to construct strings from
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > expressions when the template includes multiple interpolation sequences or a
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > mixture of literal strings and interpolations. This deprecation applies only
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > to templates that consist entirely of a single interpolation sequence.
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > (and one more similar warning elsewhere)
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > 
> 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/plan.go:67 > Will perform following changes: Add: 11, Change: 0, Destroy: 0

**Comment by ar3ndt** [comment]
> When running plan command with `--debug` flag you'll see following warning. It's worth to clear it up:
> 
> ...
> 
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Plan: 11 to add, 0 to change, 0 to destroy.
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Warning: Interpolation-only expressions are deprecated
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >   on networking.tf line 37, in resource "azurerm_subnet_network_security_group_association" "subnets_to_nsg":
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >   37:   subnet_id                 = "${azurerm_subnet.subnets[count.index].id}"
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Terraform 0.11 and earlier required all non-constant expressions to be
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > provided via interpolation syntax, but this pattern is now deprecated. To
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > silence this warning, remove the "${ sequence from the start and the }"
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > sequence from the end of this expression, leaving just the inner expression.
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > Template interpolation syntax is still used to construct strings from
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > expressions when the template includes multiple interpolation sequences or a
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > mixture of literal strings and interpolations. This deprecation applies only
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > to templates that consist entirely of a single interpolation sequence.
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 > (and one more similar warning elsewhere)
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/common.go:33 >
> > 2021-02-03T12:49:55Z INF go/src/github.com/epiphany-platform/m-azure-basic-infrastructure/cmd/plan.go:67 > Will perform following changes: Add: 11, Change: 0, Destroy: 0

done

**Review by atsikham — APPROVED** [review]

**Review by mkyc — APPROVED** [review]
LGTM

**Pull request merged**